### PR TITLE
[release/6.0] Prevent debugger deadlock with code versioning logic

### PR DESCRIPTION
**Pull request by hoyosjs**
This issue has been fixed on main, but a backport is more invasive than needed.

# Description

Various operations can result in deadlock (step in, step out, and some others) if the transitions happen to need to get to get the code versioning lock and there's a code load or a JIT happening in the background while the debugger is trying to complete a step. Each thread will lock for the other's execution operation to complete. 

# Customer Impact

Deadlocks while stepping or running under a debugger if any code versioning feature is enabled are possible - killing the process is the only solution and there's no feasible workaround for scenarios like debugger attach.

# Regression

It's a regression with respect to 3.1.

# Testing

Pending customer validation.

# Risk

Small - debugger only change. Auditing shows all callees grab the lock as needed. There are other bugs that remain around data correctness, but that change is more invasive. 

# Package authoring signed off?

N/A

### CODE REVIEW

**Review by noahfalk — APPROVED** [review]

**Review by tommcdon — APPROVED** [review]
Thank you @hoyosjs!

**Comment by carlossanlop** [comment]
@hoyosjs needs `servicing consider` label.

**Comment by hoyosjs** [comment]
It's still in draft mode, need to get customer confirmation.

**Comment by carlossanlop** [comment]
My bad 😄

**Comment by JulieLeeMSFT** [comment]
@hoyosjs it is now ready for Tactics approval. Code complete due date for servicing is May 16th. So, please get an approval over an email. I sent you an email with how to.

**Comment by JulieLeeMSFT** [comment]
@hoyosjs, I sent an email CCing you.

**Review by jeffschwMSFT — APPROVED** [review]
Approved.  We are considering for 6.0.x

**Comment by mmitche** [comment]
Reminder that code complete is today.

**Comment by jeffschwMSFT** [comment]
@carlossanlop this is ready to merge if the branch is ready.

**Pull request merged**
